### PR TITLE
Infra mvp baby

### PR DESCRIPTION
**Pull request by JaafarRammal**
# Summary

MVP ready for main

## Additional Information
• Tower
• Agents
• Simen
• Message passing
• Tower interaction

## Test Plan
See #16 for future as well as verified mutex logic for async accesses

### CODE REVIEW

File: cmd/backend/main.go [hunk 1]
```diff
@@ -1,11 +1,21 @@
 package main
 
 import (
-	"fmt"
-
-	"github.com/SOMAS2021/SOMAS2021/pkg/utils/hello"
+	simulation "github.com/SOMAS2021/SOMAS2021/pkg/simulation"
```
Comment (jzzheng22) [inline cmd/backend/main.go:4]:
```suggestion
	"github.com/SOMAS2021/SOMAS2021/pkg/simulation"
```

File: cmd/backend/main.go [hunk 2]
```diff
@@ -1,11 +1,21 @@
...
+	numOfAgents := []int{5, 5}
+	agentHP := 100
+	days := 10.0
+	reshufflePeriod := 3.0 // in days
+
+	iterationsPerDay := 24.0 * 60.0
+	iterations := int(days * iterationsPerDay)
+	iterationsBetweenReshuffles := int(reshufflePeriod * iterationsPerDay)
+
+	simEnv := simulation.NewSimEnv(foodOnPlatform, numOfAgents, agentHP, iterations, iterationsBetweenReshuffles)
```
Comment (jzzheng22) [inline cmd/backend/main.go:19]:
```suggestion
	days := 10
	daysPerReshuffle := 3

	iterationsPerDay := 24 * 60
	iterations := days * iterationsPerDay
	reshufflePeriod := daysPerReshuffle * iterationsPerDay

	simEnv := simulation.NewSimEnv(foodOnPlatform, numOfAgents, agentHP, iterations, reshufflePeriod)
```

File: go.mod [hunk 3]
```diff
@@ -1,3 +1,6 @@
 module github.com/SOMAS2021/SOMAS2021
 
 go 1.17
+
+// github.com/divan/goabm v0.0.0-20190225193149-39b1c19c3a2d
```
Comment (jzzheng22) [inline go.mod:5]:
```suggestion
```

File: pkg/agents/team1/agent1/agent.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
+package agent1
+
+import (
+	"log"
+
+	infra "github.com/SOMAS2021/SOMAS2021/pkg/infra"
```
Comment (jzzheng22) [inline pkg/agents/team1/agent1/agent.go:6]:
```suggestion
	"github.com/SOMAS2021/SOMAS2021/pkg/infra"
```

File: pkg/agents/team1/agent1/agent.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	"github.com/SOMAS2021/SOMAS2021/pkg/messages"
+	abm "github.com/SOMAS2021/SOMAS2021/pkg/utils/abm"
```
Comment (jzzheng22) [inline pkg/agents/team1/agent1/agent.go:8]:
```suggestion
	"github.com/SOMAS2021/SOMAS2021/pkg/utils/abm"
```

File: pkg/agents/team1/agent2/agent.go [hunk 6]
```diff
@@ -0,0 +1,27 @@
+package agent2
+
+import (
+	"log"
+
+	infra "github.com/SOMAS2021/SOMAS2021/pkg/infra"
+	abm "github.com/SOMAS2021/SOMAS2021/pkg/utils/abm"
```
Comment (jzzheng22) [inline pkg/agents/team1/agent2/agent.go:7]:
```suggestion
	"github.com/SOMAS2021/SOMAS2021/pkg/infra"
	"github.com/SOMAS2021/SOMAS2021/pkg/utils/abm"
```

File: pkg/infra/baseagent.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	id        string
+	hp        int
+	floor     int
+	agentType int
+	inbox     *list.List
+	tower     *Tower
+	mx        sync.RWMutex
+}
+
+func NewBaseAgent(a *abm.ABM, aType int, agentHP int, agentFloor int, id string) (*Base, error) {
```
Comment (jzzheng22) [inline pkg/infra/baseagent.go:24]:
```suggestion
func NewBaseAgent(a *abm.ABM, agentType int, agentHP int, agentFloor int, id string) (*Base, error) {
```

File: pkg/infra/baseagent.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	}
+	tower, ok := world.(*Tower)
+	if !ok {
+		return nil, errors.New("agent needs a tower world to operate")
+	}
+	return &Base{
+		id:        id,
+		hp:        agentHP,
+		floor:     agentFloor,
+		agentType: aType,
```
Comment (jzzheng22) [inline pkg/infra/baseagent.go:37]:
```suggestion
		agentType: agentType,
```

File: pkg/infra/baseagent.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+func (a *Base) TakeFood(amountOfFood float64) float64 {
+	if uint64(a.floor) == a.tower.currPlatFloor {
+		foodTaken := math.Min(a.tower.currPlatFood, amountOfFood)
+		a.updateHP(foodTaken)
+		a.tower.currPlatFood -= foodTaken
+		return foodTaken
+	}
+	return 0.0
+	// return a.tower.FoodRequest(a.id, amountOfFood)
```
Comment (jzzheng22) [inline pkg/infra/baseagent.go:85]:
```suggestion
```

File: pkg/infra/tower.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+	"log"
+	"math/rand"
+
+	"github.com/SOMAS2021/SOMAS2021/pkg/messages"
+)
+
+type Tower struct {
+	currPlatFood    float64
+	maxPlatFood     float64
+	currPlatFloor   uint64
```
Comment (jzzheng22) [inline pkg/infra/tower.go:13]:
```suggestion
	currPlatFloor   int
```

File: pkg/infra/tower.go [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+	agentCount      int
+	agents          map[string]*Base
+	agentsPerFloor  int
+	ticksPerDay     int
+	missingAgents   map[int][]int // key: floor, value: types of missing agents
+	reshufflePeriod int
+	tickCounter     int
+}
+
+func NewTower(currPlatFood float64, currPlatFloor uint64, agentCount,
```
Comment (jzzheng22) [inline pkg/infra/tower.go:23]:
```suggestion
func NewTower(currPlatFood float64, currPlatFloor, agentCount,
```

File: pkg/infra/tower.go [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+func (t *Tower) Tick() {
+	//logs
+	log.Printf("A log from the tower! Tick no: %d", t.tickCounter)
+	log.Printf("The food left on the platform = %f", t.currPlatFood)
+
+	//useful parameters
+	day := 24 * 60
+	numOfFloors := t.agentCount / int(t.agentsPerFloor)
```
Comment (jzzheng22) [inline pkg/infra/tower.go:47]:
```suggestion
	numOfFloors := t.agentCount / t.agentsPerFloor
```

File: pkg/infra/tower.go [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+	platformMovePeriod := day / numOfFloors // can add min/max
+
+	if (t.tickCounter)%(t.reshufflePeriod) == 0 {
```
Comment (jzzheng22) [inline pkg/infra/tower.go:50]:
```suggestion
        // Shuffle the agents
	if t.tickCounter % t.reshufflePeriod == 0 {
```

File: pkg/infra/tower.go [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+		t.reshuffle(numOfFloors)
+	}
+	if (t.tickCounter)%(platformMovePeriod) == 0 {
```
Comment (jzzheng22) [inline pkg/infra/tower.go:53]:
```suggestion
        // Move the platform
	if t.tickCounter % platformMovePeriod == 0 {
```

File: pkg/infra/tower.go [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+		t.currPlatFloor++
+	}
+	if (t.tickCounter)%(day) == 0 {
+		t.hpDecay() // deacreases HP and kills if < 0
+		t.ResetTower()
+	}
```
Comment (jzzheng22) [inline pkg/infra/tower.go:59]:
```suggestion
        // Decrease agent HP and reset tower at end of day
	if t.tickCounter % day == 0 {
		t.hpDecay() // decreases HP and kills if < 0
		t.ResetTower()
	}
```

File: pkg/infra/tower.go [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+
+	t.tickCounter++
+}
+
+func (t *Tower) GetMissingAgents() map[int][]int {
```
Comment (jzzheng22) [inline pkg/infra/tower.go:64]:
```suggestion
func (t *Tower) UpdateMissingAgents() map[int][]int {
```

File: pkg/infra/tower.go [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+		} else {
+			agent.setHP(newHP)
+		}
+	}
+}
+
+func (t *Tower) SendMessage(direction int, senderFloor int, msg messages.Message) {
+	log.Printf("tower sending message")
+	for _, agent := range t.agents {
+		if agent.floor == senderFloor+direction {
```
Comment (jzzheng22) [inline pkg/infra/tower.go:110]:
```suggestion
		if agent.floor == senderFloor + direction {
```

File: pkg/simulation/simulation.go [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+			sE.createNewAgent(a, t, i, agentIndex)
+			agentIndex++
+		}
+	}
+	a.LimitIterations(sE.Iterations)
+	sE.simulationLoop(a, t)
+}
+
+func (sE *SimEnv) simulationLoop(a *abm.ABM, t *infra.Tower) {
+
```
Comment (jzzheng22) [inline pkg/simulation/simulation.go:55]:
```suggestion
```

File: pkg/infra/baseagent.go [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+func (a *Base) setHP(newHP int) {
+	a.hp = newHP
+}
+
+func (a *Base) updateHP(foodTaken float64) {
+	a.hp = int(math.Min(100, float64(a.hp)+foodTaken))
+}
+
+func (a *Base) TakeFood(amountOfFood float64) float64 {
+	if uint64(a.floor) == a.tower.currPlatFloor {
```
Comment (tomeaton17) [inline pkg/infra/baseagent.go:78]: Need to be careful with this cast, make sure that a.floor is not negative (which it shouldn't be)

File: pkg/utils/abm/abm.go [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+	"log"
+	"sort"
+	"sync"
+)
+
+type ABM struct {
+	mx     sync.RWMutex
+	agents []Agent
+
+	i     int // current iteration
```
Comment (jzzheng22) [inline pkg/utils/abm/abm.go:13]:
```suggestion
	currIteration     int
```

File: pkg/utils/abm/abm.go [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+func (a *ABM) SetReportFunc(fn func(*ABM)) {
+	a.reportFunc = fn
+}
+
+func (a *ABM) AddAgent(agent Agent) {
+	a.mx.Lock()
+	a.agents = append(a.agents, agent)
+	a.mx.Unlock()
```
Comment (jzzheng22) [inline pkg/utils/abm/abm.go:45]:
```suggestion
	a.mx.Lock()
	defer a.mx.Unlock()
	a.agents = append(a.agents, agent)
```

File: pkg/utils/abm/abm.go [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+			a.agents = []Agent{}
+		} else {
+			a.agents = a.agents[1:]
+		}
+	} else if index == len(a.agents)-1 {
+		a.agents = a.agents[:len(a.agents)-1]
+	} else {
+		a.agents = append(a.agents[:index], a.agents[index+1:]...)
+	}
+	a.mx.Unlock()
```
Comment (jzzheng22) [inline pkg/utils/abm/abm.go:62]:
```suggestion
	a.mx.Lock()
	defer a.mx.Unlock()
	if index == 0 {
		if len(a.agents) == 1 {
			a.agents = []Agent{}
		} else {
			a.agents = a.agents[1:]
		}
	} else if index == len(a.agents)-1 {
		a.agents = a.agents[:len(a.agents)-1]
	} else {
		a.agents = append(a.agents[:index], a.agents[index+1:]...)
	}
```

File: pkg/utils/abm/abm.go [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+	for i := 0; i < n; i++ {
+		agent := spawnFunc(a)
+		a.AddAgent(agent)
+	}
+}
+
+func (a *ABM) LimitIterations(n int) {
+	a.mx.Lock()
+	a.limit = n
+	a.mx.Unlock()
```
Comment (jzzheng22) [inline pkg/utils/abm/abm.go:75]:
```suggestion
	a.mx.Lock()
	defer a.mx.Unlock()
	a.limit = n
```

File: pkg/utils/abm/abm.go [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+
+// Iteration returns current iteration (age, generation).
+func (a *ABM) Iteration() int {
+	a.mx.RLock()
+	defer a.mx.RUnlock()
+	return a.i
+}
+
+func (a *ABM) SimulationIterate(i int) {
+
```
Comment (jzzheng22) [inline pkg/utils/abm/abm.go:92]:
```suggestion
```

File: pkg/infra/tower.go (hunk 12 above)
Comment (tomeaton17) [inline pkg/infra/tower.go:47]: In the future, I think that the number of floors should be chosen at instantiation and independent of number of agents

File: pkg/utils/abm/abm.go [hunk 25]
```diff
@@ -0,0 +1,147 @@
...
+	a.mx.RLock()
+	defer a.mx.RUnlock()
+	return a.limit
+}
+
+// Iteration returns current iteration (age, generation).
+func (a *ABM) Iteration() int {
+	a.mx.RLock()
+	defer a.mx.RUnlock()
+	return a.i
```
Comment (jzzheng22) [inline pkg/utils/abm/abm.go:88]:
```suggestion
	return a.currIteration
```

File: pkg/utils/abm/abm.go [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+func (a *ABM) SimulationIterate(i int) {
+
+	agentsToRemove := []int{}
+	a.i = i
```
Comment (jzzheng22) [inline pkg/utils/abm/abm.go:94]:
```suggestion
	a.currIteration = i
```

File: pkg/infra/tower.go [hunk 27]
```diff
@@ -0,0 +1,121 @@
...
+		}
+		agent.setFloor(newFloor + 1)
+		remainingVacancies[newFloor]--
+	}
+}
+
+func (t *Tower) hpDecay() {
+	// TODO: can add a parameter
+	for _, agent := range t.agents {
+		newHP := agent.HP() - 20
```
Comment (tomeaton17) [inline pkg/infra/tower.go:96]: Parametrise this decay

File: pkg/infra/baseagent.go (hunk 19 above)
Comment (jzzheng22) [inline pkg/infra/baseagent.go:78]:
```suggestion
	if a.floor == a.tower.currPlatFloor {
```

**Review by jzzheng22 — CHANGES_REQUESTED** [review]
Thanks @woonmoon @udai-arneja @JaafarRammal @Silvertrousers for all of your work here

A few style changes to make but after that this is good for mvp

File: pkg/infra/tower.go (hunk 12 above)
Comment (JaafarRammal) [reply pkg/infra/tower.go:47]: Can be parametrized 👍 will be changed in the future and can be PRed by agent teams

File: pkg/infra/tower.go (hunk 27 above)
Comment (JaafarRammal) [reply pkg/infra/tower.go:96]: Yeah that's why there is the TODO 😄 Feel free to make an issue for it! Agent teams can PR on top of this function

File: pkg/infra/baseagent.go (hunk 19 above)
Comment (JaafarRammal) [reply pkg/infra/baseagent.go:78]: Fixed, thanks

File: pkg/infra/tower.go (hunk 27 above)
Comment (tomeaton17) [reply pkg/infra/tower.go:96]: Apologies, skipped over that comment.

**Review by jzzheng22 — APPROVED** [review]
I've only looked at this briefly since I'm on my phone but LGTM

**Pull request merged**